adapter-pinecone/src/main/java/io/nosqlbench/adapter/pinecone/opdispensers/PineconeUpsertOpDispenser.java

```diff
@@ -0,0 +1,128 @@
+ * limitations under the License.
+ */
+
+package io.nosqlbench.adapter.pinecone.opdispensers;
+
+import com.google.protobuf.Struct;
+import com.google.protobuf.Value;
+import io.nosqlbench.adapter.pinecone.PineconeDriverAdapter;
+import io.nosqlbench.adapter.pinecone.PineconeSpace;
+import io.nosqlbench.adapter.pinecone.ops.PineconeOp;
+import io.nosqlbench.adapter.pinecone.ops.PineconeUpsertOp;
+import io.nosqlbench.engine.api.templating.ParsedOp;
+import io.pinecone.proto.SparseValues;
+import io.pinecone.proto.UpsertRequest;
+import io.pinecone.proto.Vector;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+
+import java.util.*;
+import java.util.function.BiConsumer;
+import java.util.function.LongFunction;
+
+public class PineconeUpsertOpDispenser extends PineconeOpDispenser {
+    private static final Logger LOGGER = LogManager.getLogger(PineconeUpsertOpDispenser.class);
+    private final LongFunction<UpsertRequest.Builder> upsertRequestFunc;
+    private final LongFunction<Collection<Vector>> upsertVectorFunc;
+
+    /**
+     * Create a new PineconeUpsertOpDispenser subclassed from {@link PineconeOpDispenser}.
+     *
+     * @param adapter           The associated {@link PineconeDriverAdapter}
+     * @param op                The {@link ParsedOp} encapsulating the activity for this cycle
+     * @param pcFunction        A function to return the associated context of this activity (see {@link PineconeSpace})
+     * @param targetFunction    A LongFunction that returns the specified Pinecone Index for this Op
+     */
+    public PineconeUpsertOpDispenser(PineconeDriverAdapter adapter,
+                                     ParsedOp op,
+                                     LongFunction<PineconeSpace> pcFunction,
+                                     LongFunction<String> targetFunction) {
+        super(adapter, op, pcFunction, targetFunction);
+        upsertRequestFunc = createUpsertRequestFunc(op);
+        upsertVectorFunc = createUpsertRequestVectorsFunc(op);
+    }
+
+    private LongFunction<Collection<Vector>> createUpsertRequestVectorsFunc(ParsedOp op) {
+        Optional<LongFunction<List>> baseFunc =
+            op.getAsOptionalFunction("upsert_vectors", List.class);
+        return baseFunc.<LongFunction<Collection<Vector>>>map(listLongFunction -> l -> {
+            List<Vector> returnVectors = new ArrayList<>();
+            List<Map<String, Object>> vectors = listLongFunction.apply(l);
+            for (Map<String, Object> vector : vectors) {
+                Vector.Builder vb = Vector.newBuilder();
+                String[] rawValues = ((String) vector.get("values")).split(",");
+                ArrayList<Float> floatValues = new ArrayList<>();
+                for (String val : rawValues) {
+                    floatValues.add(Float.valueOf(val));
+                }
+                vb.addAllValues(floatValues);
+                if (vector.containsKey("sparse_values")) {
```

Comment:
It seems that there may be an opportunity here to centralize some type wrapping or tranformation logic.